### PR TITLE
fix: [SQLSRV] Query Builder always sets "<database>"."<schema>". to the table name.

### PR DESCRIPTION
**Pull request by ping-yee**
**Description**
See #8697 

**Checklist:**
- [x] Securely signed commits
- [ ] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

File: system/Database/SQLSRV/Builder.php [hunk 1]
```diff
@@ -290,6 +290,15 @@ private function getFullName(string $table): string
         }
 
         if ($this->db->escapeChar === '"') {
+            if (str_contains($table, '.')) {
+                $dbInfo    = explode('.', $table);
+                $database  = str_replace('"', '', $dbInfo[0]);
+                $schema    = str_replace('"', '', $dbInfo[1]);
+                $tableName = str_replace('"', '', $dbInfo[2]);
```
Comment (samsonasik) [inline system/Database/SQLSRV/Builder.php:297]: probably validate first if `$dbInfo` index 1 and 2 exists first?

Comment (ping-yee) [reply system/Database/SQLSRV/Builder.php]: Done.

File: system/Database/SQLSRV/Builder.php [hunk 2]
```diff
@@ -290,6 +290,21 @@ private function getFullName(string $table): string
         }
 
         if ($this->db->escapeChar === '"') {
+            if (str_contains($table, '.')) {
```
Comment (samsonasik) [inline system/Database/SQLSRV/Builder.php:293]: it may need to ensure "." is not in first and last char check as well so it always on between chars

Comment (samsonasik) [reply system/Database/SQLSRV/Builder.php:293]:
```suggestion
            if (str_contains($table, '.') && ! str_starts_with($table, '.') && ! str_ends_with($table, '.')) {
```

File: system/Database/SQLSRV/Builder.php [hunk 3]
```diff
@@ -290,6 +290,21 @@ private function getFullName(string $table): string
...
+                $schema   = $this->db->schema;
+                $table    = $dbInfo[0];
+
+                if (isset($dbInfo[1], $dbInfo[2])) {
+                    $database  = str_replace('"', '', $dbInfo[0]);
+                    $schema    = str_replace('"', '', $dbInfo[1]);
+                    $tableName = str_replace('"', '', $dbInfo[2]);
+                }
+
+                return '"' . $database . '"."' . $schema . '"."' . str_replace('"', '', $tableName) . '"' . $alias;
```
Comment (samsonasik) [inline system/Database/SQLSRV/Builder.php:305]:
```suggestion
                return '"' . $database . '"."' . $schema . '"."' . str_replace('"', '', $tableName) . '"' . $alias;
                 }
```

File: system/Database/SQLSRV/Builder.php [hunk 4]
```diff
@@ -290,6 +290,21 @@ private function getFullName(string $table): string
         }
 
         if ($this->db->escapeChar === '"') {
+            if (str_contains($table, '.')) {
+                $dbInfo   = explode('.', $table);
+                $database = $this->db->getDatabase();
+                $schema   = $this->db->schema;
```
Comment (samsonasik) [inline system/Database/SQLSRV/Builder.php:296]:
```suggestion
```

File: system/Database/SQLSRV/Builder.php (hunk 2 above)
Comment (ping-yee) [reply system/Database/SQLSRV/Builder.php]: Done.

File: system/Database/SQLSRV/Builder.php (hunk 3 above)
Comment (ping-yee) [reply system/Database/SQLSRV/Builder.php]: Done.

File: tests/system/Database/Builder/FromTest.php [hunk 5]
```diff
@@ -153,4 +153,18 @@ public function testFromSubqueryWithSQLSRV(): void
...
     }
+
+    /**
+     * @see https://github.com/codeigniter4/CodeIgniter4/issues/8697
+     */
+    public function testConstructorWithMultipleSegmentTableWithSQLSRV(): void
+    {
+        $this->db = new MockConnection(['DBDriver' => 'SQLSRV', 'database' => 'test', 'schema' => 'dbo']);
+
+        $builder = new SQLSRVBuilder('database.dbo.table', $this->db);
```
Comment (kenjis) [inline tests/system/Database/Builder/FromTest.php:164]: What if `dbo.table`?

File: system/Database/SQLSRV/Builder.php [hunk 6]
```diff
@@ -290,6 +290,20 @@ private function getFullName(string $table): string
         }
 
         if ($this->db->escapeChar === '"') {
+            if (str_contains($table, '.') && ! str_starts_with($table, '.') && ! str_ends_with($table, '.')) {
+                $dbInfo   = explode('.', $table);
+                $database = $this->db->getDatabase();
+                $table    = $dbInfo[0];
+
+                if (isset($dbInfo[1], $dbInfo[2])) {
```
Comment (samsonasik) [inline system/Database/SQLSRV/Builder.php:298]: only verify index 2 seems should be enough, explode always get 2 data at least for contain:

```suggestion
                if (isset($dbInfo[2])) {
```


ref https://3v4l.org/6YBtM

File: tests/system/Database/Builder/FromTest.php (hunk 5 above)
Comment (ping-yee) [reply tests/system/Database/Builder/FromTest.php:164]: I will extend the code to add this scenario.

Comment (ping-yee) [reply tests/system/Database/Builder/FromTest.php:164]: Done.

File: system/Database/SQLSRV/Builder.php (hunk 6 above)
Comment (ping-yee) [reply system/Database/SQLSRV/Builder.php]: Done.

File: system/Database/SQLSRV/Builder.php [hunk 7]
```diff
@@ -295,10 +295,15 @@ private function getFullName(string $table): string
...
-                    $tableName = str_replace('"', '', $dbInfo[2]);
+                if (count($dbInfo) >= 2) {
+                    if (count($dbInfo) === 3) {
+                        $database  = str_replace('"', '', $dbInfo[0]);
+                        $schema    = str_replace('"', '', $dbInfo[1]);
+                        $tableName = str_replace('"', '', $dbInfo[2]);
+                    } else {
+                        $schema    = str_replace('"', '', $dbInfo[0]);
+                        $tableName = str_replace('"', '', $dbInfo[1]);
+                    }
```
Comment (samsonasik) [inline system/Database/SQLSRV/Builder.php]: it always at least 2, so direct check 3 should be ok:

```php
                if (count($dbInfo) === 3) {
                    $database  = str_replace('"', '', $dbInfo[0]);
                    $schema    = str_replace('"', '', $dbInfo[1]);
                    $tableName = str_replace('"', '', $dbInfo[2]);
                } else {
                    $schema    = str_replace('"', '', $dbInfo[0]);
                    $tableName = str_replace('"', '', $dbInfo[1]);
                }
```

Comment (ping-yee) [reply system/Database/SQLSRV/Builder.php]: Done.

**Review by samsonasik — APPROVED** [review]

**Comment by kenjis** [comment]
@codeigniter4/database-team 
This PR resolves the issue consistent with the current implementation. 
However, such ad hoc code complicates the code and is not a fundamental solution.

Then I am not sure why only SQLSRV generates table names with database and schema names.
I am not sure if this PR should be merged.

**Comment by sclubricants** [comment]
Do we have this problem with postgres too? I know it uses schema sometimes as well..

**Comment by kenjis** [comment]
@sclubricants No. Only SQLSRV has the `getFullName()` method and use it.
https://github.com/codeigniter4/CodeIgniter4/blob/971b0da7dd3275ac8e2ba93094b0195f1b455607/system/Database/SQLSRV/Builder.php#L282

This test passes.
```diff
--- a/tests/system/Database/Builder/FromTest.php
+++ b/tests/system/Database/Builder/FromTest.php
@@ -15,6 +15,7 @@ namespace CodeIgniter\Database\Builder;
 
 use CodeIgniter\Database\BaseBuilder;
 use CodeIgniter\Database\SQLSRV\Builder as SQLSRVBuilder;
+use CodeIgniter\Database\Postgre\Builder as PostgreBuilder;
 use CodeIgniter\Test\CIUnitTestCase;
 use CodeIgniter\Test\Mock\MockConnection;
 
@@ -153,4 +154,15 @@ final class FromTest extends CIUnitTestCase
 
         $this->assertSame($expectedSQL, str_replace("\n", ' ', $builder->getCompiledSelect()));
     }
+
+    public function testConstructorWithMultipleSegmentTable(): void
+    {
+        $this->db = new MockConnection(['DBDriver' => 'Postgre', 'database' => 'test', 'schema' => 'dbo']);
+
+        $builder = new PostgreBuilder('database.dbo.table', $this->db);
+
+        $expectedSQL = 'SELECT * FROM "database"."dbo"."table"';
+
+        $this->assertSame($expectedSQL, str_replace("\n", ' ', $builder->getCompiledSelect()));
+    }
 }
```

**Comment by michalsn** [comment]
@kenjis

> Then I am not sure why only SQLSRV generates table names with database and schema names.
I am not sure if this PR should be merged.

It was related to some reported issue, but I can't find it now.

I believe it was something related to using non-default `schema`. In that case, the generated queries weren't using the correct one...

AFAIR in SQLSRV you can't declare something like `use mySchemaName`. So if we're using the non-default schema, we have to declare it precisely every time in the query.

**Comment by kenjis** [comment]
@michalsn This? #4246

**Comment by kenjis** [comment]
> AFAIR in SQLSRV you can't declare something like `use mySchemaName`. So if we're using the non-default schema, we have to declare it precisely every time in the query.

I googled, but yes, there seems no way to set schema in a connection.
https://stackoverflow.com/questions/4942023/set-default-schema-for-a-sql-query

**Comment by michalsn** [comment]
Yes, that was the issue/PR I had in mind.

The `getFullName()` method was probably created to simplify the code. I've never worked with non-default schema, so I can't determine the exact differences with CI4, but it seems to me that in CI3 we had to manually declare the schema if it wasn't the default one. In CI4 we have a config variable for that, so implementing the `getFullName()` was a convenient way to build the proper table name.

**Comment by kenjis** [comment]
Okay, in summary, it looks like this?

- in CI3, we cannot set `schema` in DB config, and CI3 does nothing to schema in gererated SQL statements.
- in CI4, we can set `schema` in DB config, and there is no way to set the schema in a DB connection. So CI4 generates full tablename (database.schema.table)

**Review by kenjis — APPROVED** [review]

**Comment by michalsn** [comment]
Yes.

**Review by michalsn — APPROVED** [review]

**Comment by kenjis** [comment]
@ping-yee Thank you!

**Pull request merged**